include/aws/crt/io/Stream.h

```diff
@@ -16,21 +16,110 @@
 
 #include <aws/crt/Exports.h>
 #include <aws/crt/Types.h>
-
-struct aws_input_stream;
+#include <aws/io/stream.h>
 
 namespace Aws
 {
     namespace Crt
     {
         namespace Io
         {
-            /**
-             * Factory to create a aws-c-io input stream subclass from a C++ stream
+            using StreamStatus = aws_stream_status;
+            using OffsetType = aws_off_t;
+
+            enum class SeekBasis
+            {
+                Begin = AWS_SSB_BEGIN,
+                End = AWS_SSB_END,
+            };
+
+            /***
+             * Interface for building an Object oriented stream that will be honored by the CRT's low-level
+             * aws_input_stream interface. To use, create a subclass of InputStream and define the abstract
+             * functions.
              */
-            AWS_CRT_CPP_API aws_input_stream *AwsInputStreamNewCpp(
-                const std::shared_ptr<Aws::Crt::Io::IStream> &stream,
-                Aws::Crt::Allocator *allocator = DefaultAllocator()) noexcept;
+            class AWS_CRT_CPP_API InputStream
+            {
+              public:
+                virtual ~InputStream();
+
+                InputStream(const InputStream &) = delete;
+                InputStream &operator=(const InputStream &) = delete;
+                InputStream(InputStream &&) = delete;
+                InputStream &operator=(InputStream &&) = delete;
+
+                explicit operator bool() const noexcept { return IsGood(); }
+
+                virtual bool IsGood() const noexcept;
```

Comment:
Well, the base class has to do setup stuff, how about I just leave m_good protected and unmark the function as virtual?

Reply:
The base class can't fail during setup anymore.
I'd personally leave it all up to the implementation, to avoid confusion over when it is/isn't automagically set by the base class, but that's just me.

Reply:
that's fine, good point. will update